### PR TITLE
Ids in TlaDecl and refactoring of TransformationTracker

### PR DESCRIPTION
**Pull request by konnov**
This is refactoring toward #262.

Most importantly,
  - `TlaDecl` is now `Identifiable`, which means that declarations have unique ids now.
  - `TransformationTracker` is now tracking transformations of declarations.

Tracking of declaration changes is a bit ugly, though it is done the same way as tracking of expression transformations.

### CODE REVIEW

File: tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala [hunk 1]
```diff
@@ -25,4 +25,34 @@ class SaveToStoreTracker(sourceStore: SourceStore) extends TransformationTracker
...
+    * Decorate a declaration transformation with a tracker.
+    *
+    * @param tr a declaration transformation
+    * @return a new declaration transformation that applies tr and tracks the changes
+    */
+  override def trackDecl(tr: TlaDeclTransformation): TlaDeclTransformation = {
+    input: TlaDecl =>
+      val output = tr(input)
+      if (output.ID != input.ID) {
+        sourceStore.findOrLog(input.ID).foreach { loc => sourceStore.add(output.ID, loc) }
```
Comment (Kukovec) [inline tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala:39]: Looking at this particular line, it wasn't immediately obvious what `findOrLog` did, I would suggest changing the name to `findWithLogging`

File: tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala [hunk 2]
```diff
@@ -25,4 +25,34 @@ class SaveToStoreTracker(sourceStore: SourceStore) extends TransformationTracker
...
+      output
+  }
+
+  /**
+    * A specialized version of trackDecl, which is tuned to TlaOperDecl.
+    *
+    * @param tr a declaration transformation
+    * @return a new declaration transformation that applies tr and tracks the changes
+    */
+  override def trackOperDecl(tr: TlaOperDecl => TlaOperDecl): TlaOperDecl => TlaOperDecl = {
```
Comment (Kukovec) [inline tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala:50]: Why is there a need for a specialized method like this? `trackDecl` should work in all cases where `trackOperDecl` is used.

File: tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/impl/TrackerWithListeners.scala [hunk 3]
```diff
@@ -8,15 +9,43 @@ import at.forsyte.apalache.tla.lir.transformations._
...
+    newDecl
+  }
+
+  /**
+    * A specialized version of trackDecl, which is tuned to TlaOperDecl.
+    *
+    * @param transformation a declaration transformation
+    * @return a new declaration transformation that applies tr and tracks the changes
+    */
+  override def trackOperDecl(
```
Comment (Kukovec) [inline tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/impl/TrackerWithListeners.scala:44]: Same as above, please explain the need for `TlaOperDecl`-specific methods.

File: tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/package.scala [hunk 4]
```diff
@@ -1,11 +1,38 @@
...
+    * @return a declaration transformation that is tracking updates to declarations and expressions inside them.
+    */
+  def fromExToDeclTransformation(
+      tracker: TransformationTracker,
+      transformation: TlaExTransformation
+  ): TlaDeclTransformation = tracker.trackDecl {
+    case d @ TlaOperDecl(_, _, b) =>
+      val copy = d.copy(body = transformation(b))
+      // see issue #345
+      copy.isRecursive = d.isRecursive
```
Comment (Kukovec) [inline tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/package.scala:30]: Unnecessary, the `.copy` method on `TlaOperDecl` is not the default one and correctly copies `.isRecursive`.

File: tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/Flatten.scala [hunk 5]
```diff
@@ -50,11 +50,14 @@ object Flatten {
...
-          )
-        }
+        def xform: TlaOperDecl => TlaOperDecl =
+          tracker.trackOperDecl { d: TlaOperDecl =>
+            d.copy(
+              body = self(d.body)
+            )
+          }
+
+        val newDefs = defs map xform
```
Comment (Kukovec) [inline tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/Flatten.scala:60]: You could have just written `defs map (tracker.trackOperDecl { ... })`

**Review by Kukovec — APPROVED** [review]
Looks good overall, though I fear we might later find that we're missing declaration tracking in some `case LetInEx(...) => ...`.

**Comment by Kukovec** [comment]
@konnov could you please address this? I'm working on #436 , which would benefit from having tracking on declarations.

**Comment by konnov** [comment]
Yep, I will work on it tomorrow

File: tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala (hunk 1 above)
Comment (konnov) [reply tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala:39]: renamed to `findOrWarn`

File: tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala (hunk 2 above)
Comment (konnov) [reply tla-import/src/main/scala/at/forsyte/apalache/tla/imp/src/SaveToStoreTracker.scala:50]: just for convenience, as `TlaOperDecl` is by far the most often used transformation.

File: tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/impl/TrackerWithListeners.scala (hunk 3 above)
Comment (konnov) [reply tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/impl/TrackerWithListeners.scala:44]: Added an explanation to TransformationTracker

File: tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/Flatten.scala (hunk 5 above)
Comment (konnov) [reply tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/Flatten.scala:60]: Agree. This is less verbose:  `defs map tracker.trackOperDecl { d => d.copy(body = self(d.body)) }`. I sometimes forget how easy it is to stack functions in Scala :)

**Pull request merged**